### ISSUE TITLE
Way to utilize `loss_weights` in custom `fit` method!

### ISSUE THREAD

**Issue by innat** [keras-team/keras#17260]
I'm going through this documents ([customizing_what_happens_in_fit](https://keras.io/guides/customizing_what_happens_in_fit/) by @fchollet), and eventually feel like getting stuck if I want to weighted the loss value. For example, observe the following code

```python
class CustomModel(keras.Model):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.loss_tracker = keras.metrics.Mean(name="loss")
        self.mae_metric = keras.metrics.MeanAbsoluteError(name="mae")

    def train_step(self, data):
        x, y = data

        with tf.GradientTape() as tape:
            y_pred = self(x, training=True)  # Forward pass
            # Compute our own loss
            loss = keras.losses.mean_squared_error(y, y_pred)

        # Compute gradients
        trainable_vars = self.trainable_variables
        gradients = tape.gradient(loss, trainable_vars)

        # Update weights
        self.optimizer.apply_gradients(zip(gradients, trainable_vars))

        # Compute our own metrics
        self.loss_tracker.update_state(loss)
        self.mae_metric.update_state(y, y_pred)
        return {"loss": self.loss_tracker.result(), "mae": self.mae_metric.result()}

    @property
    def metrics(self):
        # We list our `Metric` objects here so that `reset_states()` can be
        # called automatically at the start of each epoch
        # or at the start of `evaluate()`.
        # If you don't implement this property, you have to call
        # `reset_states()` yourself at the time of your choosing.
        return [self.loss_tracker, self.mae_metric]
```
```python
# Construct an instance of CustomModel
inputs = keras.Input(shape=(32,))
outputs = keras.layers.Dense(1)(inputs)
model = CustomModel(inputs, outputs)

# We don't passs a loss or metrics here.
model.compile(optimizer="adam")

# Just use `fit` as usual -- you can use callbacks, etc.
x = np.random.random((1000, 32))
y = np.random.random((1000, 1))
model.fit(x, y, epochs=5)
```

In the above, no loss is passed in `model.compile` method, because, we're manually computing loss value inside the `train_step` method. That's great. Now, let's assume I've more than one output from model and want to weighted them towards the final loss. In such case, what would be the approach to do so?

In `model.compile`, it has `loss_weights` arguments that can be used but for above cases, how to achieve this?

```python
Model.compile(
    optimizer="rmsprop",
    loss=None,
    metrics=None,
    loss_weights=None,
    weighted_metrics=None,
    run_eagerly=None,
    steps_per_execution=None,
    jit_compile=None,
    **kwargs
)
```

In the above mention [document](https://keras.io/guides/customizing_what_happens_in_fit/), it somewhat discuss about `class_weights` and `sample_weights`, but not about `loss_weights`.

**Comment by innat** [comment]
**additional info**

There is a class function [`compute_loss` ](https://www.tensorflow.org/api_docs/python/tf/keras/Model#compute_loss) of `keras.Model`, it states

![image](https://user-images.githubusercontent.com/17668390/202859984-abe76a9e-9661-4044-9863-7db4b04e5beb.png)

Above, is the `sample_weight` argument right option here to weighted the loss? Is it computationally same as `loss_weights` found in the `.compile` method? If so,

- how to use it for single or multi-output loss calculation, how to use it? any sample code?
- what if, I want to change the `sample_weights` value in the middle of training, how to do that?
- About name, why naming `sample_weights` instead of `loss_weights`? Is it a bit confusing?

**Comment by tilakrayal** [comment]
@innat,
We would prefer a PR for the validation, Could you please feel free to submit a PR for the requested change to be made. Thank you!

**Comment by innat** [comment]
@tilakrayal 
Thanks for the response.  I'm not open for contribution right now, please find someone else. Also. could you please elaborate more about [`compute_loss`](https://www.tensorflow.org/api_docs/python/tf/keras/Model#compute_loss) class method? It might be the right function here. Currently, it gives 

```python
compute_loss(
    x=None, y=None, y_pred=None, sample_weight=None
)
```
whereas I think it should be 

```python
compute_loss(
    x=None, y=None, y_pred=None, loss_weights=None
)
```

where, for single output or multi-output,

````python
len(y) == len(y_pred) == len(loss_weights)
len(y.keys()) == len(y_pred.keys()) == len(loss_weights.keys())
````

```python
Model.compile(
    optimizer="rmsprop",
    loss=None,
    metrics=None,
    loss_weights=None,
)
```

Here are two things, (**one**), if we complie the loss function inside the `.compile` method we can pass `loss_weights`; (**two**), what if we don't compile the loss function, will it be still possible to use `loss_weights` inside the `.compile` method and retrievable to [`compute_loss`](https://www.tensorflow.org/api_docs/python/tf/keras/Model#compute_loss) class method?

---

This makes a follow-up question about `metrics` and `weighted metrics` params.

```python
Model.compile(
    optimizer="rmsprop",
    metrics=None,
    weighted_metrics=None,
```

**Comment by rchao** [comment]
@innat, is it an option for your use case to invoke multiple loss functions, collect the losses, multiply them by the weights you would like, sum them, and feed that final combined weighted loss value to `tape.gradient()` api?

**Comment by innat** [comment]
@rchao 
Thanks for the response. About your question, yes, sort of.

- The main query explained in the first post. How to weight the custom loss function in case of custom fit. Suppose that, we won't pass `loss` argument in the `model.compile` method. So, in such scenario, how to use `loss_weights` for custom loss function. Please see the code example in the first post for more details.
- [Follow up Q.] Both cases single output and multiple output model!

Also, I'm curious to know aobut compute_loss, details [here](https://github.com/keras-team/keras/issues/17260#issuecomment-1322022913). What is the role it can play here?

---

Eventually, I've just realized that, I can pass initial weights value to `keras.Model` instance and using callback, I can access them ( and update them also ) to weights the loss function inside the `tf.GradientTape` context.  Would you please validate the approach or is there any the better way?

```python
class CustomModel(keras.Model):
    def __init__(self, weights, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.weights = weights
        self.loss_tracker = keras.metrics.Mean(name="loss")
        self.mae_metric = keras.metrics.MeanAbsoluteError(name="mae")

   def train_step(...):
       # use self.weights to anything we want to ...
```

**Comment by rchao** [comment]
Is there anything actionable or we can close this issue?

**Comment by innat** [comment]
Yes, it can be closed. Thank you.

**Comment by rchao** [comment]
Thanks Mohammed!
